bothan-lib/src/worker/rest.rs

```diff
@@ -45,30 +47,58 @@ pub async fn start_polling<S: Store, E: Display, P: AssetInfoProvider<Error = E>
 
         select! {
             _ = cancellation_token.cancelled() => break,
-            r = timeout(interval.period(), provider.get_asset_info(&ids)) => handle_polling_result(r, &store).await,
+            r = poll_with_timeout(&provider, &ids, update_interval, metrics.clone()) => handle_polling_result(r, &store, metrics.clone()).await,
         }
     }
 }
 
+async fn poll_with_timeout<P, E>(
+    provider: &P,
+    ids: &[String],
+    timeout_interval: Duration,
+    metrics: RestMetrics,
+) -> Result<Result<Vec<AssetInfo>, E>, Elapsed>
+where
+    E: Display,
+    P: AssetInfoProvider<Error = E>,
+{
+    let start_time = chrono::Utc::now().timestamp_millis();
+    let result: Result<Result<Vec<AssetInfo>, E>, Elapsed> =
+        timeout(timeout_interval, provider.get_asset_info(ids)).await;
+    let elapsed_time = chrono::Utc::now().timestamp_millis() - start_time;
+
+    match result {
+        Ok(Ok(_)) => metrics.record_polling_duration(elapsed_time, PollingResult::Success),
+        Ok(Err(_)) => metrics.record_polling_duration(elapsed_time, PollingResult::Failed),
+        Err(_) => metrics.record_polling_duration(elapsed_time, PollingResult::Timeout),
+    }
+
+    result
+}
+
 async fn handle_polling_result<S, E>(
     poll_result: Result<Result<Vec<AssetInfo>, E>, Elapsed>,
     store: &WorkerStore<S>,
+    metrics: RestMetrics,
 ) where
     S: Store,
     E: Display,
 {
     match poll_result {
         Ok(Ok(asset_info)) => {
+            metrics.increment_polling_total(PollingResult::Success);
             if let Err(e) = store.set_batch_asset_info(asset_info).await {
                 error!("failed to update asset info with error: {e}");
             } else {
                 info!("asset info updated successfully");
             }
         }
         Ok(Err(e)) => {
+            metrics.increment_polling_total(PollingResult::Failed);
             error!("failed to update asset info with error: {e}");
         }
         Err(_) => {
+            metrics.increment_polling_total(PollingResult::Timeout);
             error!("updating interval exceeded timeout");
         }
     }
```

Comment:
use `let match` to get the `PollingResult` then increment outside the match instead